src/cpp/common/alts_context.cc

```diff
@@ -0,0 +1,112 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <grpc/grpc_security.h>
+#include <grpcpp/alts_context.h>
+
+#include "src/core/lib/gprpp/memory.h"
+#include "src/core/tsi/alts/handshaker/alts_tsi_handshaker.h"
+#include "src/cpp/common/secure_auth_context.h"
+#include "src/proto/grpc/gcp/altscontext.upb.h"
+
+namespace grpc {
+
+AltsContext::AltsContext(const grpc_gcp_AltsContext* ctx) {
+  upb_strview application_protocol =
+      grpc_gcp_AltsContext_application_protocol(ctx);
+  application_protocol_ =
+      std::string(application_protocol.data, application_protocol.size);
+  upb_strview record_protocol = grpc_gcp_AltsContext_record_protocol(ctx);
+  record_protocol_ = std::string(record_protocol.data, record_protocol.size);
+  upb_strview peer_service_account =
+      grpc_gcp_AltsContext_peer_service_account(ctx);
+  peer_service_account_ =
+      std::string(peer_service_account.data, peer_service_account.size);
+  upb_strview local_service_account =
+      grpc_gcp_AltsContext_local_service_account(ctx);
+  local_service_account_ =
+      std::string(local_service_account.data, local_service_account.size);
+  const grpc_gcp_RpcProtocolVersions* versions =
+      grpc_gcp_AltsContext_peer_rpc_versions(ctx);
+  if (versions != nullptr) {
+    const grpc_gcp_RpcProtocolVersions_Version* max_versions =
+        grpc_gcp_RpcProtocolVersions_max_rpc_version(versions);
+    if (max_versions != nullptr) {
+      int max_version_major =
+          grpc_gcp_RpcProtocolVersions_Version_major(max_versions);
+      int max_version_minor =
+          grpc_gcp_RpcProtocolVersions_Version_minor(max_versions);
+      peer_rpc_versions_.max_rpc_versions.major_version = max_version_major;
+      peer_rpc_versions_.max_rpc_versions.minor_version = max_version_minor;
+    }
+    const grpc_gcp_RpcProtocolVersions_Version* min_versions =
+        grpc_gcp_RpcProtocolVersions_min_rpc_version(versions);
+    if (min_versions != nullptr) {
+      int min_version_major =
+          grpc_gcp_RpcProtocolVersions_Version_major(min_versions);
+      int min_version_minor =
+          grpc_gcp_RpcProtocolVersions_Version_minor(min_versions);
+      peer_rpc_versions_.min_rpc_versions.major_version = min_version_major;
+      peer_rpc_versions_.min_rpc_versions.minor_version = min_version_minor;
+    }
+  }
+  security_level_ = static_cast<grpc_security_level>(
+      (int)grpc_gcp_AltsContext_security_level(ctx));
```

Comment:
Note you are casting an int to an enum which I think triggers undefined behavior if the int is out of range. Maybe you can do a function to check against the min and max levels to make sure it is in range, and if not return a `nullptr` as the context?